### ISSUE TITLE
Problem with calculation

### ISSUE THREAD

**Issue by ldir-hector** [spotlessmind1975/ugbasic#552]
I had another problem with calculation, and I got results that were wrong.
I tried to isolate the problem in a simplified program and I wrote this : 

```
BITMAP ENABLE (640,200,2)
CLS

dec_x = (SIGNED 32BIT)120
dec_y =  (SIGNED 32BIT)74

depart_x = (SIGNED 32BIT)76
depart_y =(SIGNED 32BIT) 127

departX = departX + (dec_x)/100
departY = departY - (dec_y)/100
	
dec_x=dec_x*100
dec_y=dec_y*100
	
dec_x=dec_x/125
dec_y=dec_y/125
	
dec_y=dec_y - 10
PRINT departX;" ";departY;" ";dec_x;" ";dec_y
```
And I got this : 
![image](https://user-images.githubusercontent.com/118670680/212466321-17447596-456c-4da7-8266-2eab35dfdc1b.png)

What I don't understand is that the two last results 96 and 49 are correct, but the two first are wrong, I should have 77 and 126 or maybe 127 but not 0 !

In my more complex program this code is iterated many times and on the second result depart_y I got very weird results I got -65409 instead of 0 but the first result is 77 as expected : 
![image](https://user-images.githubusercontent.com/118670680/212466493-f7a3a3e0-f66e-40ec-9359-abaab090b7f9.png)

Here is the whole code : 

```
BITMAP ENABLE (640,200,2)
CLS
DIM monCosinus AS SIGNED 32BIT(180) = #{9998,9994,9986,9976,9962,9945,9925,9903,9877,9848,9816,9781,9744,9703,9659,9613,9563,9511,9455,9397,9336,9272,9205,9135,9063,8988,8910,8829,8746,8660,8572,8480,8387,8290,8192,8090,7986,7880,7771,7660,7547,7431,7314,7193,7071,6947,6820,6691,6561,6428,6293,6157,6018,5878,5736,5592,5446,5299,5150,5000,4848,4695,4540,4384,4226,4067,3907,3746,3584,3420,3256,3090,2924,2756,2588,2419,2250,2079,1908,1736,1564,1392,1219,1045,872,698,523,349,175,0,-175,-349,-523,-698,-872,-1045,-1219,-1392,-1564,-1736,-1908,-2079,-2250,-2419,-2588,-2756,-2924,-3090,-3256,-3420,-3584,-3746,-3907,-4067,-4226,-4384,-4540,-4695,-4848,-5000,-5150,-5299,-5446,-5592,-5736,-5878,-6018,-6157,-6293,-6428,-6561,-6691,-6820,-6947,-7071,-7193,-7314,-7431,-7547,-7660,-7771,-7880,-7986,-8090,-8192,-8290,-8387,-8480,-8572,-8660,-8746,-8829,-8910,-8988,-9063,-9135,-9205,-9272,-9336,-9397,-9455,-9511,-9563,-9613,-9659,-9703,-9744,-9781,-9816,-9848,-9877,-9903,-9925,-9945,-9962,-9976,-9986,-9994,-9998,-10000}
DIM monSinus AS SIGNED 32BIT(180)= #{175,349,523,698,872,1045,1219,1392,1564,1736,1908,2079,2250,2419,2588,2756,2924,3090,3256,3420,3584,3746,3907,4067,4226,4384,4540,4695,4848,5000,5150,5299,5446,5592,5736,5878,6018,6157,6293,6428,6561,6691,6820,6947,7071,7193,7314,7431,7547,7660,7771,7880,7986,8090,8192,8290,8387,8480,8572,8660,8746,8829,8910,8988,9063,9135,9205,9272,9336,9397,9455,9511,9563,9613,9659,9703,9744,9781,9816,9848,9877,9903,9925,9945,9962,9976,9986,9994,9998,10000,9998,9994,9986,9976,9962,9945,9925,9903,9877,9848,9816,9781,9744,9703,9659,9613,9563,9511,9455,9397,9336,9272,9205,9135,9063,8988,8910,8829,8746,8660,8572,8480,8387,8290,8192,8090,7986,7880,7771,7660,7547,7431,7314,7193,7071,6947,6820,6691,6561,6428,6293,6157,6018,5878,5736,5592,5446,5299,5150,5000,4848,4695,4540,4384,4226,4067,3907,3746,3584,3420,3256,3090,2924,2756,2588,2419,2250,2079,1908,1736,1564,1392,1219,1045,872,698,523,349,175,0}

angle = (SIGNED 32BIT) 45
PRINT monCosinus(angle-1)
PRINT monSinus(angle-1)
dec_x =  monCosinus(angle-1)/4
dec_y =  monSinus(angle-1)/4 

departX = (SIGNED 32BIT) 0
departY = (SIGNED 32BIT) 200
PRINT (departX >= 0);" faux=";(2 = 3);" ";departY >= 0
 PRINT	departX;" ";departY;" ";dec_x;" ";dec_y
WAIT KEY
WHILE departY < 201 AND departX < 160 AND departX >= 0  AND departY >= 0 

	departX = departX + (dec_x)/100
	departY = departY - (dec_y)/100
	
	dec_x=dec_x*100
	dec_y=dec_y*100
	
	dec_x=dec_x/125
	dec_y=dec_y/125
	
	dec_y=dec_y - 10
	PRINT departX;" ";departY;" ";dec_x;" ";dec_y
	WAIT KEY

WEND

```
That's quite puzzling to see negative numbers passing the test departY >= 0 !

**Comment by spotlessmind1975** [comment]
Hello @ldir-hector and *than your for bug reporting!**

I think that the first strange value depends on the fact that the printed variable (`departX`) is never initialized, as is the case with the calculated variable (`depart_x`). Please, could confirm this fact?

The second example, on the other hand, highlighted a real problem in subtraction with 32-bit numbers under Z80. Long story short, the two's complement of a 32-bit number is computed as the juxtaposition of two 16-bit two's complement, but no carry was carried if the low 16-bit increment overflowed . Now it should work, but I ask you to verify. Thank you!

Fixed with COLDFIX v1.12.1 [rev. 20230114.1313].

**Comment by ldir-hector** [comment]
That's it ! I shall be more carefull naming my variables !
That makes me think about something, in Visual Basic, you can put "Option Explicit" on top of your source and then you need to declare all your variables. That might be interesting.

**Comment by ldir-hector** [comment]
But I still got a wrong result for the second program, and I got it even sooner : 
![image](https://user-images.githubusercontent.com/118670680/212471954-d6671038-5311-495c-a274-7bc0ed14944a.png)
I should got 17 183 1413 1403 like in the second screenshot.

**Comment by spotlessmind1975** [comment]
Hello @ldir-hector , **thank you for your feedback!**

I'm looking into why it's not working.

Anyway, I follow your suggestion here: #553 .

**Comment by spotlessmind1975** [comment]
Fixed with COLDFIX v1.12.1 [rev. 20230114.1443].

**Comment by ldir-hector** [comment]
I tested the last release and noticed that there is still something wrong with negative numbers : 

```
BITMAP ENABLE (640,200,2)
CLS

dec_y = (SIGNED 32BIT) 0
dec_y = -155
dec_y=dec_y*100
dec_y=dec_y/125

PRINT "Should be -124 ";dec_y
dec_y=dec_y - 200
PRINT "Should be -324 ";dec_y;
```
You got this : 
![image](https://user-images.githubusercontent.com/118670680/212492788-15f9e406-eb81-4234-9253-a8fab53541aa.png)
I don't understand why I got 400 instead of -124.
I noticed that if I put 155 instead of -155, it works fine.

**Comment by spotlessmind1975** [comment]
Hello @ldir-hector and **thank you for your bug reporting!**

The issue has been fixed on COLDFIX v1.12.1 [rev. 20230114.2225].
